src/server_manager/web_app/digitalocean_server.ts

```diff
@@ -69,105 +69,106 @@ export class DigitalOceanServer extends ShadowboxServer implements server.Manage
     // to better encapsulate the management api address logic.
     super(id);
     console.info('DigitalOceanServer created');
-    this.eventQueue.once('server-active', () => console.timeEnd('activeServer'));
-    this.pollInstallState();
-  }
-
-  waitOnInstall(): Promise<void> {
-    return new Promise((fulfill, reject) => {
-      // Poll this.installState for changes.  This can poll quickly as it
-      // will not make any network requests.
-      const intervalId = setInterval(() => {
-        if (this.installState === InstallState.UNKNOWN) {
-          // installState not known, wait until next retry.
-          return;
-        }
-        // State is now known, so we can stop checking.
-        clearInterval(intervalId);
-        if (this.installState === InstallState.SUCCESS) {
-          fulfill();
-        } else if (this.installState === InstallState.ERROR) {
-          reject(new errors.ServerInstallFailedError());
-        } else if (this.installState === InstallState.DELETED) {
-          reject(new errors.DeletedServerError());
-        }
-      }, 100);
-    });
+    if (this.isInstallCompleted()) {
+      this.setApiUrlAndCertificate();
+      this.installState = InstallState.COMPLETED;
+    }
   }
 
   // Sets this.installState, will keep polling until this.installState can
   // be set to something other than UNKNOWN.
-  private pollInstallState(): void {
+  async *installProcess(): AsyncGenerator<number, void> {
```

Comment:
This approach looks ok, but one thing that concerns me is that:

1. the install state won't be updated if this is not called, unlike the original code
2. this method is expected to be read only, but actually modifies the object state

An alternative is to move the code back to a loop you start in the constructor, emit some `install-state-changed` events, then, in the installProgress method, yield the initial progress, and then loop over the events. The EventEmitter gives you an async generator for events with the `on` method:
https://nodejs.org/api/events.html#events_events_on_emitter_eventname_options

Reply:
OK. I followed that pattern.  Unfortunately, "events.on" is not part of the [eventemitter3](https://www.npmjs.com/package/eventemitter3) package, so I wrote my own implementation.

If you prefer this approach, I'll add some unit tests.  If you'd prefer to take an external dependency, I think we could probably achieve a similar effect using [RxJS](https://rxjs.dev/) or [readable-stream](https://www.npmjs.com/package/readable-stream).

Reply:
I'm sorry about that the `eventemitter3`, I didn't realize that.
It seems RxJS doesn't use async generators and I don't think it's worth changing your implementation anyway.